### PR TITLE
feat(entity): Adds Entity, EntityCard, EntityPin, EntityVertex

### PR DESCRIPTION
**Pull request by jbrundage**
@GordonSmith please review

Signed-off-by: Jaman <jbrundage372@gmail.com>

### CODE REVIEW

**Comment by jbrundage** [comment]
image

File: packages/common/src/Entity.ts [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                    .diameter(context.annotationDiameter())
+                    .paddingPercent(context.annotationPaddingPercent())
+                    .target(this)
+                    .shape("square")
+                    ;
+                if (_d.faChar) anno.faChar(_d.faChar);
+                if (_d.shape) anno.shape(_d.shape);
+                if (_d.image_colorFill) anno.image_colorFill(_d.image_colorFill);
+                if (_d.shape_colorFill) anno.shape_colorFill(_d.shape_colorFill);
+                if (_d.shape_colorStroke) anno.shape_colorStroke(_d.shape_colorStroke);
```
Comment (GordonSmith) [inline packages/common/src/Entity.ts]: For all these tests you should be testing for undefined (especially the colors as color(0) should be black and not none).

Also they need to be able to unset the color for when they change from "red" to "":
```
anno.shape_colorStroke(_d.shape_colorStroke) 
```
is probably simpler and better?

File: packages/common/src/Entity.ts [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    protected _element_title: d3SelectionType;
+
+    constructor() {
+        super();
+
+        this._background_widget = new Shape();
+        this._icon_widget = new Icon();
+        this._title_widget = new Text();
+        this._desc_widget = new Text();
+        this._annotation_widgets = {};
```
Comment (GordonSmith) [inline packages/common/src/Entity.ts]: Should probably be: 
```javascript
this._annotation_widgets:{[idx: nunber]: {widget:Widget, bbox:{x:number, y:number, width:number, height:number}}} = {};
```

File: packages/common/src/Entity.ts [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                if (_d.shape) anno.shape(_d.shape);
+                if (_d.image_colorFill) anno.image_colorFill(_d.image_colorFill);
+                if (_d.shape_colorFill) anno.shape_colorFill(_d.shape_colorFill);
+                if (_d.shape_colorStroke) anno.shape_colorStroke(_d.shape_colorStroke);
+                context._annotation_widgets[idx] = anno;
+            })
+            ;
+        for (const i in context._annotation_widgets) {
+            context._annotation_widgets[i].render();
+            context._annotation_widgets[i]._bbox = context._annotation_widgets[i].getBBox(true);
```
Comment (GordonSmith) [inline packages/common/src/Entity.ts]: With the above typing change you won't be stuffing _bbox into a widget object,

File: packages/common/src/EntityVertex.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        let anno_w_sum = 0;
+        const anno_gap = this.annotationSpacing();
+        for (const i in this._annotation_widgets) {
+            const anno_bbox = this._annotation_widgets[i]._bbox;
+            this._annotation_widgets[i].move({
+                x: (text_bbox.width / 2) - (anno_bbox.width / 2) - anno_w_sum + anno_gap,
+                y: (text_bbox.height / 2) + (anno_bbox.height / 2) - anno_gap
+            });
+            anno_w_sum += anno_bbox.width + anno_gap;
+        }
```
Comment (GordonSmith) [inline packages/common/src/EntityVertex.ts]: You should have a common function for this which positions the row of annotations at a given place.

File: packages/common/src/EntityPin.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        let anno_w_sum = 0;
+        const anno_gap = this.annotationSpacing();
+        for (const i in this._annotation_widgets) {
+            const anno_bbox = this._annotation_widgets[i]._bbox;
+            this._annotation_widgets[i].move({
+                x: -anno_w_sum + (background_bbox.width / 2) - anno_bbox.width,
+                y: _anno_y + (anno_bbox.height / 2)
+            });
+            anno_w_sum += anno_bbox.width + anno_gap;
+        }
```
Comment (GordonSmith) [inline packages/common/src/EntityPin.ts]: You should have a common function for this which positions the row of annotations at a given place.

File: packages/common/src/EntityCard.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+        let anno_w_sum = 0;
+        for (const i of Object.keys(this._annotation_widgets)) {
+            const anno_bbox = this._annotation_widgets[i]._bbox;
+            this._annotation_widgets[i].move({
+                x: (_background_w / 2) - (anno_bbox.width / 2) - anno_w_sum - this.padding(),
+                y: (_background_h / 2) - (anno_bbox.height / 2) - this.padding()
+            });
+            anno_w_sum += anno_bbox.width + anno_gap;
+        }
```
Comment (GordonSmith) [inline packages/common/src/EntityCard.ts]: You should have a common function for this which positions the row of annotations at a given place.

File: packages/common/src/Shape.ts [hunk 7]
```diff
@@ -87,12 +87,48 @@ export class Shape extends SVGWidget {
...
         }
     }
+    getShapeElementTag(_shape) {
+        if (_shape === "square") {
+            return "rect";
+        }
+        if (_shape === "pin") {
+            return "path";
+        }
+        return _shape;
```
Comment (GordonSmith) [inline packages/common/src/Shape.ts]: Use switch statement

**Pull request merged**